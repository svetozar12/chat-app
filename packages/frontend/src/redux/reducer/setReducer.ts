import { Action, ActionType, InitialState2 } from "../state";
export const initialState = {
  reciever: "",
  chat_inviter: "",
  pageNumber: 2,
  setUserSettings: false,
  setFriendRequest: false,
  toggleCreateGroup: false,
  setMobileNav: false,
  setIsMatch: false,
  setChatSettings: false,
  setModalInvite: false,
  setIsLoggingIn: false,
};
const setReducer = (state: InitialState2 | any | InitialState2[] = initialState, action: Action) => {
  switch (action.type) {
    case ActionType.SET_RECIEVER:
      return { ...state, reciever: action.payload };
    case ActionType.CHAT_INVITER:
      return { ...state, chat_inviter: action.payload };
    case ActionType.INCREMENT_PAGE_NUMBER:
      return { ...state, pageNumber: ++action.payload };
    case ActionType.SET_FRIEND_REQUEST:
      return { ...state, setFriendRequest: action.payload };
    case ActionType.SET_USER_SETTINGS:
      return { ...state, setUserSettings: action.payload };
    case ActionType.TOGGLE_CREATE_GROUP:
      return { ...state, toggleCreateGroup: action.payload };
    case ActionType.SET_MOBILE_NAV:
      return { ...state, setMobileNav: action.payload };
    case ActionType.SET_CHAT_SETTINGS:
      return { ...state, setChatSettings: action.payload };
    case ActionType.SET_IS_MATCH:
      return { ...state, setIsMatch: action.payload };
    case ActionType.SET_MODAL_INVITE:
      return { ...state, setModalInvite: action.payload };
    case ActionType.SET_IS_LOGEDIN:
      return { ...state, setIsLoggingIn: action.payload };
    case ActionType.SIGN_OUT:
      return initialState;
    default:
      return state;
  }
};

export default setReducer;
