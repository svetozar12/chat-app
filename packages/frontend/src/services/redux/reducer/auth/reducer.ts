import { HYDRATE } from 'next-redux-wrapper';
import { IAuth } from './state';
import { ActionType } from '../../types';

export const initialState: IAuth = {
  remember_me: false,
  loginPrompt: false,
};
const authReducer = (state = initialState, action: any) => {
  switch (action.type) {
    case ActionType.REMEMBER_ME_CHECK:
      return { ...state, remember_me: action.payload };
    case ActionType.SIGN_IN:
      return { ...state, cookie: action.payload };
    default:
      return state;
  }
};

export default authReducer;
