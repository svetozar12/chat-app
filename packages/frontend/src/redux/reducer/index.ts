import { combineReducers } from "redux";
import reducer from "./authReducer/authReducer";
import setReducer from "./setReducer/setReducer";
import saveInputReducer from "./save_inputReducer/save_inputReducer";
import messageReducer from "./messageReducer/messageReducer";
import { ActionType } from "../types";
const combiReducers = combineReducers({
  authReducer: reducer,
  setReducer,
  messageReducer,
  saveInputReducer,
});

const combReducers = (state: any, action: any) => {
  if (action.type === ActionType.SIGN_OUT) {
    return undefined;
  }
  return combiReducers(state, action);
};

export default combReducers;
