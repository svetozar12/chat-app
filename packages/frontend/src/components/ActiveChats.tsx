/** @jsxRuntime classic */
/** @jsx jsx */
import { jsx } from "@emotion/react";
import React from "react";
import { useRouter } from "next/router";
import { Socket } from "socket.io-client";
import { useDispatch, useSelector } from "react-redux";
import Avatar from "./Avatar";
import styled from "@emotion/styled";
import { InitialState2 } from "../redux/state";
import { BsThreeDots } from "react-icons/bs";
import { css } from "@emotion/css";
interface IActiveChats {
  _id: string;
  members: string[];
  cookieName: string;
  socketRef: Socket;
  chatId: string;
}

const Space_between = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

const Space_between_child = styled.div`
  margin: 0;
  color: black;
  display: flex;
`;

const Contacts_info = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
`;

const Contacts_container = styled.div`
  color: var(--main-white);
  width: 97%;
  cursor: pointer;
  padding: 1rem 1rem;
  width: 100%;
  white-space: nowrap;
  transition: 0.2s;
  border-radius: 15px;
  white-space: nowrap;
  &:hover {
    border-radius: 15px;
    background: rgba(122, 122, 122, 0.1);
    transition: 0.2s;
  }
`;

const Last_message = styled.p`
  margin: 0;
  color: #65676b;
  justify-content: flex-start;
  align-items: center;
  margin: 1rem 0;
`;

const ActiveChats = ({
  _id,
  members,
  cookieName,
  socketRef,
  chatId,
}: IActiveChats) => {
  const router = useRouter();
  const [user1, user2] = [members[0], members[1]];
  const [inviter, setInviter] = React.useState<string>("");
  const dispatch = useDispatch();
  const state = useSelector(
    (state: { setReducer: InitialState2 }) => state.setReducer,
  );

  React.useEffect(() => {
    const notMe: string[] = members.filter((element) => element !== cookieName);
    setInviter(notMe[0]);
  }, []);

  const joinChat = () => {
    socketRef?.emit("join_chat", {
      chat_id: cookieName,
    });
    router.push(`${_id}`);
  };

  const dispatching = () => {
    dispatch({
      type: "TOGGLE_CREATE_GROUP",
      payload: false,
    });
  };
  const chatSettings = () => {
    dispatch({
      type: "SET_CHAT_SETTINGS",
      payload: !state.setChatSettings,
    });
  };

  return (
    <Contacts_container
      onClick={() => {
        joinChat();
        dispatching();
      }}
      className={`${
        _id === chatId &&
        css`
          border-radius: 15px;
          background: rgba(122, 122, 122, 0.1);
        `
      } `}
    >
      <Space_between>
        <section
          style={{
            display: "flex",
            alignItems: "center",
          }}
        >
          <Avatar members={members} inviter={inviter} cookieName={cookieName} />
          <Contacts_info>
            <Space_between_child>
              {members.length > 2
                ? members.map((element, index) => {
                    if (index === 3) return;
                    return (
                      <p style={{ margin: 0 }} key={index}>
                        {element}
                        {element[members.length - 1] === element[index]
                          ? " ..."
                          : ","}
                      </p>
                    );
                  })
                : (members.length === 1 && (
                    <p style={{ margin: 0 }}>{user1}</p>
                  )) ||
                  (user2 === cookieName && (
                    <p style={{ margin: 0 }}>{user1}</p>
                  )) ||
                  (user1 === cookieName && (
                    <p style={{ margin: 0 }}>{user2}</p>
                  ))}
            </Space_between_child>

            <Last_message>Last message...</Last_message>
          </Contacts_info>
        </section>
        {_id === chatId && (
          <BsThreeDots
            className={css`
              width: 2rem;
              height: 2rem;
              background: var(--main-white);
              color: var(--main-black);
              border-radius: 25px;
              box-shadow: 0 0 5px var(--main-black);
              &:hover {
                color: rgba(122, 122, 122, 1);
              }
            `}
            onClick={chatSettings}
          />
        )}
      </Space_between>
    </Contacts_container>
  );
};
export default ActiveChats;
