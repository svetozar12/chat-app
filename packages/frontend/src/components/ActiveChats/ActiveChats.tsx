import React from "react";
import { useRouter } from "next/router";
import { Socket } from "socket.io-client";
import { useDispatch, useSelector } from "react-redux";
import Avatar from "../Avatar";
import { InitialState2 } from "../../redux/state";
import { BsThreeDots } from "react-icons/bs";
import { css, cx } from "@emotion/css";
interface IActiveChats {
  _id: string;
  members: string[];
  cookieName: string;
  socketRef: Socket;
  chatId: string;
}

const ActiveChats = ({ _id, members, cookieName, socketRef, chatId }: IActiveChats) => {
  const router = useRouter();
  const [user1, user2] = [members[0], members[1]];
  const [inviter, setInviter] = React.useState<string>("");
  const dispatch = useDispatch();
  const state = useSelector((state: { setReducer: InitialState2 }) => state.setReducer);

  React.useEffect(() => {
    const notMe: string[] = members.filter((element) => element !== cookieName);
    setInviter(notMe[0]);
  }, []);

  const joinChat = () => {
    socketRef?.emit("join_chat", {
      chat_id: cookieName,
    });
    router.push(`${_id}`);
  };

  const dispatching = () => {
    dispatch({
      type: "TOGGLE_CREATE_GROUP",
      payload: false,
    });
  };
  const chatSettings = () => {
    dispatch({
      type: "SET_CHAT_SETTINGS",
      payload: !state.setChatSettings,
    });
  };

  return (
    <div
      data-testid="chat"
      onClick={() => {
        joinChat();
        dispatching();
      }}
      className={cx(
        {
          [css`
            border-radius: 15px;
            background: rgba(122, 122, 122, 0.1);
          `]: _id === chatId,
        },
        css`
          color: var(--main-white);
          width: 97%;
          cursor: pointer;
          padding: 1rem 1rem;
          width: 100%;
          white-space: nowrap;
          transition: 0.2s;
          border-radius: 15px;
          white-space: nowrap;
          &:hover {
            border-radius: 15px;
            background: rgba(122, 122, 122, 0.1);
            transition: 0.2s;
          }
        `,
      )}
    >
      <div
        className={css`
          display: flex;
          align-items: center;
          justify-content: space-between;
          width: 100%;
        `}
      >
        <section
          style={{
            display: "flex",
            alignItems: "center",
          }}
        >
          <Avatar members={members} inviter={inviter} cookieName={cookieName} />
          <div
            className={css`
              display: flex;
              flex-direction: column;
              align-items: flex-start;
              justify-content: center;
            `}
          >
            <div
              className={css`
                margin: 0;
                color: black;
                display: flex;
              `}
            >
              {members.length > 2
                ? members.map((element, index) => {
                    if (index === 3) return;
                    return (
                      <p style={{ margin: 0 }} key={index}>
                        {element}
                        {element[members.length - 1] === element[index] ? `${members.length > 3 ? "..." : ""}` : ","}
                      </p>
                    );
                  })
                : (members.length === 1 && <p style={{ margin: 0 }}>{user1}</p>) ||
                  (user2 === cookieName && <p style={{ margin: 0 }}>{user1}</p>) ||
                  (user1 === cookieName && <p style={{ margin: 0 }}>{user2}</p>)}
            </div>
            <p
              className={css`
                margin: 0;
                color: #65676b;
                justify-content: flex-start;
                align-items: center;
                margin: 1rem 0;
              `}
            >
              Last message...
            </p>
          </div>
        </section>
        {_id === chatId && (
          <BsThreeDots
            className={css`
              width: 2rem;
              height: 2rem;
              background: var(--main-white);
              color: var(--main-black);
              border-radius: 25px;
              box-shadow: 0 0 5px var(--main-black);
              &:hover {
                color: rgba(122, 122, 122, 1);
              }
            `}
            onClick={chatSettings}
          />
        )}
      </div>
    </div>
  );
};
export default ActiveChats;
