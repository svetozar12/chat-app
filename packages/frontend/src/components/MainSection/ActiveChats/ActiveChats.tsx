import React, { FC, useEffect } from 'react';
import { useRouter } from 'next/router';
import { connect } from 'react-redux';
import { css, cx } from '@emotion/css';
import { useCookie } from 'next-cookie';
import { HStack } from '@chakra-ui/react';
import { STATE } from 'services/redux/reducer';
import { bindActionCreators, Dispatch } from 'redux';
import { toggleCreateGroup } from 'services/redux/reducer/toggles/actions';
import { Chat } from 'services/generated';
import ActiveChatDetails from 'components/MainSection/ActiveChats/subcomponents/ActiveChatDetails/ActiveChatDetails';

interface IActiveChats extends Chat, ReturnType<typeof mapStateToProps>, ReturnType<typeof mapDispatchToProps> {
  chatId: string;
}

const ActiveChats: FC<IActiveChats> = ({ _id, members, ws, toggleCreateGroup }) => {
  const router = useRouter();
  const cookie = useCookie();
  const cookieName: string = cookie.get('name');
  const { acc } = router.query;
  const chatId = acc as string;

  useEffect(() => {
    joinChat();
    return () => {
      ws.ws?.off('join_chat');
    };
  }, []);

  function joinChat() {
    router.push(`${_id}`).then(() => {
      ws.ws?.emit('join_chat', {
        rooms: [cookieName, _id],
      });
    });
  }

  return (
    <HStack
      w="full"
      cursor="pointer"
      p="2rem"
      ml="4rem"
      whiteSpace="nowrap"
      transition="0.2s"
      borderRadius="2xl"
      _hover={{ borderRadius: '15px', background: ' rgba(122, 122, 122, 0.1)', transition: '0.2s' }}
      onClick={() => {
        joinChat();
        toggleCreateGroup(false);
      }}
      className={cx({
        [css`
          border-radius: 15px;
          background: rgba(122, 122, 122, 0.1);
        `]: _id === chatId,
      })}
    >
      <ActiveChatDetails _id={_id} chatId={chatId} members={members} />
    </HStack>
  );
};

const mapStateToProps = (state: STATE) => ({
  ws: state.ws,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  toggleCreateGroup: bindActionCreators(toggleCreateGroup, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(ActiveChats);
