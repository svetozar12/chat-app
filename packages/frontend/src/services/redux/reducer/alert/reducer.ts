import { AnyAction } from 'redux';
import { IAlert } from 'services/redux/reducer/alert/state';
import { ActionType } from '../../types';

const initialState: IAlert = {
  good: '', // good and bad stand for good alert and bad alert
  bad: '',
};

const alertReducer = (state = initialState, action: AnyAction) => {
  switch (action.type) {
    case ActionType.LOGIN_POST_ERROR:
      return { ...state, bad: action.payload };
    case ActionType.REGISTER_POST:
      return { ...state, good: action.payload };
    case ActionType.REGISTER_POST_ERROR:
      return { ...state, bad: action.payload };
    default:
      return state;
  }
};

export default alertReducer;
