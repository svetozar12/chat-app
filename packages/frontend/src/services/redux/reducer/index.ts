import { AnyAction, combineReducers } from 'redux';
import authReducer from 'services/redux/reducer/auth/reducer';
import { ActionType } from '../types';
import { HYDRATE } from 'next-redux-wrapper';
import wsReducer from 'services/redux/reducer/websocket/reducer';
import messageReducer from 'services/redux/reducer/messages/reducer';

import { IAuth } from 'services/redux/reducer/auth/state';
import { IWebSocket } from 'services/redux/reducer/websocket/state';
import { IMessage } from 'services/redux/reducer/messages/state';
import IInputs from 'services/redux/reducer/inputs/state';
import inputReducer from 'services/redux/reducer/inputs/reducer';
import toggleReducer from 'services/redux/reducer/toggles/reducer';
import { IToggle } from 'services/redux/reducer/toggles/state';
import IInvite from 'services/redux/reducer/invites/state';
import inviteReducer from 'services/redux/reducer/invites/reducer';
import { IAlert } from 'services/redux/reducer/alert/state';
import alertReducer from 'services/redux/reducer/alert/reducer';

export interface STATE {
  auth: IAuth;
  ws: IWebSocket;
  messages: IMessage;
  inputs: IInputs;
  toggle: IToggle;
  invite: IInvite;
  alert: IAlert;
}

const combiReducers = combineReducers({
  auth: authReducer,
  ws: wsReducer,
  messages: messageReducer,
  inputs: inputReducer,
  toggle: toggleReducer,
  invite: inviteReducer,
  alert: alertReducer,
});

const combReducers = (state: any, action: AnyAction) => {
  if (action.type === HYDRATE)
    return {
      ...state,
      ...action.payload,
    };
  if (action.type === ActionType.SIGN_OUT) {
    return undefined;
  }
  return combiReducers(state, action);
};

export default combReducers;
