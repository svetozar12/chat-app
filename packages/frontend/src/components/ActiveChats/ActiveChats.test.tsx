import ActiveChats from "./ActiveChats";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { initialState } from "../../redux/reducer/setReducer";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";

const socketRef: any = jest.fn();
const dumy: string[] = ["ivan", "gerg"];
const mockStore = configureStore([]);
const store = mockStore({
  setReducer: initialState,
});

const setupRender = () => {
  const component = render(
    <Provider store={store}>
      <ActiveChats
        _id="61c4957b735b579e5442dfe8"
        chatId="61c4957b735b579e5442dfe8"
        members={dumy}
        cookieName="ivan"
        socketRef={socketRef}
      />
    </Provider>,
  );
  return component;
};

describe("<ActiveChats/>", () => {
  beforeEach(() => {
    setupRender();
  });
  it("should create snapshot", () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <ActiveChats
              _id="61c4957b735b579e5442dfe8"
              chatId="61c4957b735b579e5442dfe8"
              members={dumy}
              cookieName="ivan"
              socketRef={socketRef}
            />
          </Provider>,
        )
        .toJSON(),
    ).toMatchSnapshot();
  });
  it("should render", () => {
    render(
      <Provider store={store}>
        <ActiveChats
          _id="61c4957b735b579e5442dfe8"
          chatId="61c4957b735b579e5442dfe8"
          members={dumy}
          cookieName="ivan"
          socketRef={socketRef}
        />
      </Provider>,
    );
    expect(screen.getAllByTitle("ivan")[0]).toBeInTheDocument();
  });
});
