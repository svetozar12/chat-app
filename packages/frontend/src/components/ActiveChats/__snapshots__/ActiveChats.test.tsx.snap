// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActiveChats/> should create snapshot 1`] = `
<div
  className="css-gzd0uh"
  data-testid="chat"
  onClick={[Function]}
>
  <div
    className="css-1lymioy"
  >
    <section
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <section
        data-testid="avatar"
        title="ivan"
      >
        <div>
          <div
            style={
              Object {
                "marginRight": "1rem",
              }
            }
          >
            <svg
              className="css-nkhozk css-boia29"
              data-testid="single_avatar"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                  "height": undefined,
                  "width": undefined,
                }
              }
              viewBox="0 0 496 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z"
              />
            </svg>
          </div>
        </div>
      </section>
      <div
        className="css-1c5u6r2"
      >
        <div
          className="css-19dbfyd"
        >
          <p
            style={
              Object {
                "margin": 0,
              }
            }
          >
            gerg
          </p>
        </div>
        <p
          className="css-1p5u1lq"
        >
          Last message...
        </p>
      </div>
    </section>
    <svg
      className="css-4t0f8h"
      fill="currentColor"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 9.5a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm5 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm5 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
      />
    </svg>
  </div>
</div>
`;
